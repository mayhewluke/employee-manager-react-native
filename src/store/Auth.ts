import firebase from "firebase";
import { NavigationScreenProp } from "react-navigation";
import { Dispatch, Reducer } from "redux";
import { action as createAction } from "typesafe-actions";

export enum AuthActionType {
  EMAIL_CHANGED = "EMAIL_CHANGED",
  PASSWORD_CHANGED = "PASSWORD_CHANGED",
  LOGIN_SUCCESS = "LOGIN_SUCCESS",
  LOGIN_FAIL = "LOGIN_FAIL",
  LOGIN_START = "LOGIN_START",
}

export interface IAuthState {
  email: string;
  password: string;
  loading: boolean;
  error: string;
  user: firebase.auth.UserCredential | null;
}

const INITIAL_STATE: IAuthState = {
  email: "",
  error: "",
  loading: false,
  password: "",
  user: null,
};

export type AuthAction =
  | { type: AuthActionType.EMAIL_CHANGED; payload: string }
  | { type: AuthActionType.PASSWORD_CHANGED; payload: string }
  | {
      type: AuthActionType.LOGIN_SUCCESS;
      payload: firebase.auth.UserCredential;
    }
  | { type: AuthActionType.LOGIN_FAIL }
  | { type: AuthActionType.LOGIN_START };

type AuthDispatch = Dispatch<AuthAction>;

export const authReducer: Reducer<IAuthState, AuthAction> = (
  state = INITIAL_STATE,
  action,
) => {
  switch (action.type) {
    case AuthActionType.EMAIL_CHANGED:
      return { ...state, email: action.payload };
    case AuthActionType.PASSWORD_CHANGED:
      return { ...state, password: action.payload };
    case AuthActionType.LOGIN_START:
      return { ...state, loading: true, error: "" };
    case AuthActionType.LOGIN_SUCCESS:
      return { ...state, ...INITIAL_STATE, user: action.payload };
    case AuthActionType.LOGIN_FAIL:
      return { ...state, loading: false, error: "Something went wrong!" };
    default:
      return state;
  }
};

const loginSuccess = (dispatch: AuthDispatch, navigate: any) => (
  user: firebase.auth.UserCredential,
) => {
  dispatch(createAction(AuthActionType.LOGIN_SUCCESS, user));
  navigate("Main");
};

const loginFail = (dispatch: AuthDispatch) => () =>
  dispatch(createAction(AuthActionType.LOGIN_FAIL));

export const emailChanged = (email: string) =>
  createAction(AuthActionType.EMAIL_CHANGED, email);

export const passwordChanged = (password: string) =>
  createAction(AuthActionType.PASSWORD_CHANGED, password);

export const logIn = (
  email: string,
  password: string,
  navigation: NavigationScreenProp<any>,
) => (dispatch: AuthDispatch) => {
  dispatch(createAction(AuthActionType.LOGIN_START));
  firebase
    .auth()
    .signInWithEmailAndPassword(email, password)
    .then(loginSuccess(dispatch, navigation.navigate))
    .catch(() =>
      firebase
        .auth()
        .createUserWithEmailAndPassword(email, password)
        .then(loginSuccess(dispatch, navigation.navigate))
        .catch(loginFail(dispatch)),
    );
};
