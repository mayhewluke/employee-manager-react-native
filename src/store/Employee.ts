import firebase from "firebase";
import { NavigationScreenProp } from "react-navigation";
import { Dispatch, Reducer } from "redux";
import { action as createAction } from "typesafe-actions";

import { Async } from "./common";

export enum EmployeeActionType {
  UPDATE_FIELD = "UPDATE_FIELD",
  CREATE_START = "CREATE_START",
  CREATE_SUCCESS = "CREATE_SUCCESS",
  CREATE_FAIL = "CREATE_FAIL",
  UPDATE_START = "UPDATE_START",
  UPDATE_SUCCESS = "UPDATE_SUCCESS",
  UPDATE_FAIL = "UPDATE_FAIL",
  EDIT = "EDIT",
  RESET = "RESET",
  SHOW_MODAL = "SHOW_MODAL",
  CLOSE_MODAL = "CLOSE_MODAL",
  FIRE_START = "FIRE_START",
  FIRE_SUCCESS = "FIRE_SUCCESS",
  FIRE_FAIL = "FIRE_FAIL",
}

export enum ShiftDay {
  Monday = "Monday",
  Tuesday = "Tuesday",
  Wednesday = "Wednesday",
  Thursday = "Thursday",
  Friday = "Friday",
  Saturday = "Saturday",
  Sunday = "Sunday",
}

export interface IEmployee<T extends null | string> {
  employeeName: string;
  phone: string;
  shift: ShiftDay;
  uid: T;
}

export interface IEmployeeState extends IEmployee<null | string> {
  createAction: Async<null>;
  updateAction: Async<null>;
  fireAction: Async<null>;
  fireModalShown: boolean;
}

export const INITIAL_STATE: IEmployeeState = {
  createAction: { state: "INIT" },
  employeeName: "",
  fireAction: { state: "INIT" },
  fireModalShown: false,
  phone: "",
  shift: ShiftDay.Monday,
  uid: null,
  updateAction: { state: "INIT" },
};

export type FieldUpdatePayload =
  | { field: "employeeName" | "phone"; value: string }
  | { field: "shift"; value: ShiftDay };

export type EmployeeAction =
  | {
      type: EmployeeActionType.UPDATE_FIELD;
      payload: FieldUpdatePayload;
    }
  | { type: EmployeeActionType.CREATE_START }
  | { type: EmployeeActionType.CREATE_SUCCESS }
  | { type: EmployeeActionType.CREATE_FAIL }
  | { type: EmployeeActionType.UPDATE_START }
  | { type: EmployeeActionType.UPDATE_SUCCESS }
  | { type: EmployeeActionType.UPDATE_FAIL }
  | { type: EmployeeActionType.EDIT; payload: IEmployee<string> }
  | { type: EmployeeActionType.RESET }
  | { type: EmployeeActionType.SHOW_MODAL }
  | { type: EmployeeActionType.CLOSE_MODAL }
  | { type: EmployeeActionType.FIRE_START }
  | { type: EmployeeActionType.FIRE_SUCCESS }
  | { type: EmployeeActionType.FIRE_FAIL };

type EmployeeDispatch = Dispatch<EmployeeAction>;

export const employeeReducer: Reducer<IEmployeeState, EmployeeAction> = (
  state = INITIAL_STATE,
  action,
) => {
  switch (action.type) {
    case EmployeeActionType.UPDATE_FIELD:
      return { ...state, [action.payload.field]: action.payload.value };
    case EmployeeActionType.CREATE_START:
      return { ...state, createAction: { state: "PROGRESS" } };
    case EmployeeActionType.CREATE_SUCCESS:
      return INITIAL_STATE;
    case EmployeeActionType.CREATE_FAIL:
      return {
        ...state,
        createAction: { state: "ERROR", error: "Something went wrong!" },
      };
    case EmployeeActionType.UPDATE_START:
      return { ...state, updateAction: { state: "PROGRESS" } };
    case EmployeeActionType.UPDATE_SUCCESS:
      return INITIAL_STATE;
    case EmployeeActionType.UPDATE_FAIL:
      return {
        ...state,
        updateAction: { state: "ERROR", error: "Something went wrong!" },
      };
    case EmployeeActionType.EDIT:
      return { ...state, ...action.payload };
    case EmployeeActionType.RESET:
      return INITIAL_STATE;
    case EmployeeActionType.SHOW_MODAL:
      return { ...state, fireModalShown: true };
    case EmployeeActionType.CLOSE_MODAL:
      return { ...state, fireModalShown: false };
    case EmployeeActionType.FIRE_START:
      return {
        ...state,
        fireAction: { state: "PROGRESS" },
        fireModalShown: false,
      };
    case EmployeeActionType.FIRE_SUCCESS:
      return INITIAL_STATE;
    case EmployeeActionType.FIRE_FAIL:
      return {
        ...state,
        fireAction: { state: "ERROR", error: "Something went wrong!" },
      };
    default:
      return state;
  }
};

export const updateField = (payload: FieldUpdatePayload) =>
  createAction(EmployeeActionType.UPDATE_FIELD, payload);

const createSuccess = (dispatch: EmployeeDispatch, navigate: any) => () => {
  navigate("EmployeeList");
  dispatch(createAction(EmployeeActionType.CREATE_SUCCESS));
};

const createFail = (dispatch: EmployeeDispatch) => () =>
  dispatch(createAction(EmployeeActionType.CREATE_FAIL));

export const createEmployee = (
  employee: IEmployee<null>,
  navigation: NavigationScreenProp<any>,
) => (dispatch: EmployeeDispatch) => {
  const { currentUser } = firebase.auth();
  if (currentUser === null) {
    navigation.navigate("Auth");
  } else {
    dispatch(createAction(EmployeeActionType.CREATE_START));
    return (
      firebase
        .database()
        .ref(`/users/${currentUser.uid}/employees`)
        .push(employee)
        // Have to put error handler within `.then` due to Firebase idiosyncracity
        .then(
          createSuccess(dispatch, navigation.navigate),
          createFail(dispatch),
        )
    );
  }
  return Promise.reject();
};

export const editEmployee = (
  navigation: NavigationScreenProp<any>,
  employee: IEmployee<string>,
) => (dispatch: EmployeeDispatch) => {
  dispatch(createAction(EmployeeActionType.EDIT, employee));
  navigation.navigate("EditEmployee", { employeeName: employee.employeeName });
};

export const resetForm = () => createAction(EmployeeActionType.RESET);

const updateSuccess = (
  dispatch: EmployeeDispatch,
  navigation: NavigationScreenProp<any>,
) => () => {
  navigation.navigate("EmployeeList");
  dispatch(createAction(EmployeeActionType.UPDATE_SUCCESS));
};

const updateFail = (dispatch: EmployeeDispatch) => () => {
  dispatch(createAction(EmployeeActionType.UPDATE_FAIL));
};

export const updateEmployee = (
  { uid, ...employee }: IEmployee<null | string>,
  navigation: NavigationScreenProp<any>,
) => (dispatch: EmployeeDispatch) => {
  const { currentUser } = firebase.auth();
  if (currentUser === null) {
    navigation.navigate("Auth");
  } else if (uid === null) {
    // TODO handle this better, either by making it impossible to get here (via
    // type safety) or showing an error message at least
    navigation.navigate("EmployeeList");
  } else {
    dispatch(createAction(EmployeeActionType.UPDATE_START));
    return firebase
      .database()
      .ref(`/users/${currentUser.uid}/employees/${uid}`)
      .set(employee)
      .then(updateSuccess(dispatch, navigation), updateFail(dispatch));
  }
  return Promise.reject();
};

const fireSuccess = (
  dispatch: EmployeeDispatch,
  navigation: NavigationScreenProp<any>,
) => () => {
  navigation.navigate("EmployeeList");
  dispatch(createAction(EmployeeActionType.FIRE_SUCCESS));
};

const fireFail = (dispatch: EmployeeDispatch) => () =>
  dispatch(createAction(EmployeeActionType.FIRE_FAIL));

export const showFireModal = () => createAction(EmployeeActionType.SHOW_MODAL);
export const closeFireModal = () =>
  createAction(EmployeeActionType.CLOSE_MODAL);

export const fireEmployee = (
  uid: string | null,
  navigation: NavigationScreenProp<any>,
) => (dispatch: EmployeeDispatch) => {
  const { currentUser } = firebase.auth();
  if (currentUser === null) {
    navigation.navigate("Auth");
  } else if (uid === null) {
    // TODO handle this better, either by making it impossible to get here (via
    // type safety) or showing an error message at least
    navigation.navigate("EmployeeList");
  } else {
    dispatch(createAction(EmployeeActionType.FIRE_START));
    return firebase
      .database()
      .ref(`/users/${currentUser.uid}/employees/${uid}`)
      .remove()
      .then(fireSuccess(dispatch, navigation), fireFail(dispatch));
  }
  return Promise.reject();
};
