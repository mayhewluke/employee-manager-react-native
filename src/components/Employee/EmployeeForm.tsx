import * as React from "react";
import { Picker, StyleSheet, Text, View } from "react-native";
import { connect } from "react-redux";

import { IRootState } from "../../store";
import {
  FieldUpdatePayload,
  ShiftDay,
  updateField,
} from "../../store/Employee";
import { CardSection, Input } from "../common";

const styles = StyleSheet.create({
  errorTextStyle: {
    alignSelf: "center",
    color: "red",
    fontSize: 20,
  },
  pickerContainerStyle: {
    flexDirection: "column",
    paddingLeft: 25,
  },
  pickerLabelStyle: {
    fontSize: 18,
  },
});

export interface IFormProps {
  employeeName: string;
  phone: string;
  shift: ShiftDay;
  dispatchFieldUpdate: (payload: FieldUpdatePayload) => void;
  error: string;
}

const Form = ({
  employeeName,
  phone,
  shift,
  dispatchFieldUpdate,
  error,
}: IFormProps) => {
  const dispatchUpdateName = (value: string) =>
    dispatchFieldUpdate({ value, field: "employeeName" });
  const dispatchUpdatePhone = (value: string) =>
    dispatchFieldUpdate({ value, field: "phone" });
  const dispatchUpdateShift = (value: ShiftDay) =>
    dispatchFieldUpdate({ value, field: "shift" });
  return (
    <View>
      <CardSection>
        <Input
          label="Name"
          placeholder="Taylor"
          value={employeeName}
          onChangeText={dispatchUpdateName}
        />
      </CardSection>
      <CardSection>
        <Input
          label="Phone"
          placeholder="555-555-5555"
          value={phone}
          onChangeText={dispatchUpdatePhone}
        />
      </CardSection>

      <CardSection style={styles.pickerContainerStyle}>
        <Text style={styles.pickerLabelStyle}>Shift day</Text>
        <Picker selectedValue={shift} onValueChange={dispatchUpdateShift}>
          {Object.keys(ShiftDay).map(key => (
            <Picker.Item
              key={key}
              value={key}
              label={ShiftDay[key as keyof typeof ShiftDay]}
            />
          ))}
        </Picker>
      </CardSection>

      {error !== "" && (
        <CardSection>
          <Text style={styles.errorTextStyle}>{error}</Text>
        </CardSection>
      )}
    </View>
  );
};

const mapStateToProps = (state: IRootState) => {
  const { employeeName, phone, shift } = state.employee;
  return { employeeName, phone, shift };
};

export default connect(
  mapStateToProps,
  { dispatchFieldUpdate: updateField },
)(Form);
