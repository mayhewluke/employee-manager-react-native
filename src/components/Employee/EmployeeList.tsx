import React, { Component } from "react";
import {
  FlatList,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from "react-native";
import { NavigationScreenProp, NavigationScreenProps } from "react-navigation";
import { connect } from "react-redux";

import { Async, IRootState } from "../../store";
import { editEmployee, IEmployee } from "../../store/Employee";
import { unwatchEmployees, watchEmployees } from "../../store/Employees";

import { CardSection, Spinner, SpinnerSize } from "../common";

const addEmployee = (navigation: NavigationScreenProp<any>) => () =>
  navigation.navigate("CreateEmployee");

const styles = StyleSheet.create({
  addEmployeeTextStyle: {
    alignSelf: "center",
    color: "#007aff",
    fontSize: 16,
    fontWeight: "600",
    paddingRight: 10,
  },
});

interface IProps extends NavigationScreenProps {
  dispatchWatchEmployees: (navigation: NavigationScreenProp<any>) => void;
  dispatchEditEmployee: any;
  employeesAction: Async<{ [uid: string]: IEmployee<string> }>;
}

class EmployeeList extends Component<IProps> {
  public static navigationOptions = ({
    navigation,
  }: {
    navigation: NavigationScreenProp<any>;
  }) => ({
    // TODO extract this into a reusable component
    headerRight: (
      <TouchableOpacity onPress={addEmployee(navigation)}>
        <Text style={styles.addEmployeeTextStyle}>Create employee</Text>
      </TouchableOpacity>
    ),
    title: "Employees",
  });

  public componentDidMount() {
    const { dispatchWatchEmployees, navigation } = this.props;
    dispatchWatchEmployees(navigation);
  }

  public componentWillUnmount() {
    unwatchEmployees();
  }

  private renderEmployee = ({
    item: employee,
  }: {
    item: IEmployee<string>;
  }) => {
    const { dispatchEditEmployee, navigation } = this.props;
    const onPress = () => dispatchEditEmployee(employee, navigation);
    return (
      <TouchableOpacity onPress={onPress}>
        <CardSection>
          <Text>{employee.employeeName}</Text>
        </CardSection>
      </TouchableOpacity>
    );
  };

  public render() {
    const { employeesAction } = this.props;
    const keyExtractor = ({ uid }: IEmployee<string>) => uid;
    return (
      <View>
        {employeesAction.state === "PROGRESS" && (
          <Spinner size={SpinnerSize.Large} />
        )}
        {employeesAction.state === "COMPLETE" && (
          <FlatList
            data={Object.keys(employeesAction.value).map((uid: string) => ({
              uid,
              ...employeesAction.value[uid],
            }))}
            renderItem={this.renderEmployee}
            keyExtractor={keyExtractor}
          />
        )}
      </View>
    );
  }
}

const mapStateToProps = (state: IRootState) => {
  const { employeesAction } = state.employees;
  return {
    employeesAction,
  };
};

export default connect(
  mapStateToProps,
  {
    dispatchEditEmployee: editEmployee,
    dispatchWatchEmployees: watchEmployees,
  },
)(EmployeeList);
