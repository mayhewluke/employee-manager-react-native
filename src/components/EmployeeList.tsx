import React, { Component } from "react";
import {
  FlatList,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from "react-native";
import { NavigationScreenProp, NavigationScreenProps } from "react-navigation";
import { connect } from "react-redux";

import { IRootState } from "../store";
import { editEmployee, IEmployee } from "../store/Employee";
import { unwatchEmployees, watchEmployees } from "../store/Employees";

import { CardSection, Spinner, SpinnerSize } from "./common";

const addEmployee = (navigation: NavigationScreenProp<any>) => () =>
  navigation.navigate("CreateEmployee");

const styles = StyleSheet.create({
  addEmployeeTextStyle: {
    alignSelf: "center",
    color: "#007aff",
    fontSize: 16,
    fontWeight: "600",
    paddingRight: 10,
  },
});

interface IProps extends NavigationScreenProps {
  dispatchWatchEmployees: (navigation: NavigationScreenProp<any>) => void;
  dispatchEditEmployee: any;
  employees: { [uid: string]: IEmployee };
  loading: boolean;
}

class EmployeeList extends Component<IProps> {
  public static navigationOptions = ({
    navigation,
  }: {
    navigation: NavigationScreenProp<any>;
  }) => ({
    // TODO extract this into a reusable component
    headerRight: (
      <TouchableOpacity onPress={addEmployee(navigation)}>
        <Text style={styles.addEmployeeTextStyle}>Create employee</Text>
      </TouchableOpacity>
    ),
    title: "Employees",
  });

  public componentDidMount() {
    const { dispatchWatchEmployees, navigation } = this.props;
    dispatchWatchEmployees(navigation);
  }

  public componentWillUnmount() {
    const { navigation } = this.props;
    unwatchEmployees(navigation);
  }

  private renderEmployee = ({ item: employee }: { item: IEmployee }) => {
    const { dispatchEditEmployee, navigation } = this.props;
    const onPress = () => dispatchEditEmployee(navigation, employee);
    return (
      <TouchableOpacity onPress={onPress}>
        <CardSection>
          <Text>{employee.employeeName}</Text>
        </CardSection>
      </TouchableOpacity>
    );
  };

  public render() {
    const { loading, employees } = this.props;
    return (
      <View>
        {loading === true && <Spinner size={SpinnerSize.Large} />}
        <FlatList
          data={Object.keys(employees).map((uid: string) => ({
            uid,
            key: uid, // For React element list indexing
            ...employees[uid],
          }))}
          renderItem={this.renderEmployee}
        />
      </View>
    );
  }
}

const mapStateToProps = (state: IRootState) => {
  const { loading, employees } = state.employees;
  return {
    employees,
    loading,
  };
};

export default connect(
  mapStateToProps,
  {
    dispatchEditEmployee: editEmployee,
    dispatchWatchEmployees: watchEmployees,
  },
)(EmployeeList);

// export default EmployeeList;
