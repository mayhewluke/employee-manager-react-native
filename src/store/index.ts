import { applyMiddleware, combineReducers, createStore, Store } from "redux";
import { combineEpics, createEpicMiddleware } from "redux-observable";
import reduxThunk from "redux-thunk";

import { AuthAction, authReducer, IAuthState } from "./Auth";
import {
  createEmployeeEpic,
  employeeReducer,
  fireEmployeeEpic,
  IEmployeeState,
  updateEmployeeEpic,
} from "./Employee";
import {
  employeesReducer,
  employeesSubscriptionEpic,
  IEmployeesState,
} from "./Employees";

const rootReducer = combineReducers({
  auth: authReducer,
  employee: employeeReducer,
  employees: employeesReducer,
});

const rootEpic = combineEpics(
  createEmployeeEpic,
  fireEmployeeEpic,
  updateEmployeeEpic,
  employeesSubscriptionEpic,
);

const epicMiddleware = createEpicMiddleware();

export interface IRootState {
  auth: IAuthState;
  employee: IEmployeeState;
  employees: IEmployeesState;
}

export const store: Store<IRootState, RootAction> = createStore(
  rootReducer,
  {},
  applyMiddleware(reduxThunk, epicMiddleware),
);

epicMiddleware.run(rootEpic);

export type RootAction = AuthAction;
export * from "./common/Async";
