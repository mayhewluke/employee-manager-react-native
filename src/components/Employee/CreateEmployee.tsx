import React, { Component } from "react";
import { NavigationScreenProp, NavigationScreenProps } from "react-navigation";
import { connect } from "react-redux";

import { Async, IRootState } from "../../store";
import {
  createEmployee,
  IEmployee,
  resetForm,
  ShiftDay,
  updateField,
} from "../../store/Employee";
import { Button, Card, CardSection, Spinner, SpinnerSize } from "../common";
import EmployeeForm, { IFormProps } from "./EmployeeForm";

// TODO extract into a reusable component
const renderButton = (isLoading: boolean, onSubmit: () => void) => {
  if (isLoading) {
    return <Spinner size={SpinnerSize.Large} />;
  }
  return <Button onPress={onSubmit}>Create</Button>;
};

interface IProps extends IFormProps, NavigationScreenProps {
  employeeName: string;
  phone: string;
  shift: ShiftDay;
  createAction: Async<null>;
  dispatchCreateEmployee: (
    employee: IEmployee<null>,
    navigation: NavigationScreenProp<any>,
  ) => void;
  dispatchResetForm: () => void;
}

class CreateEmployee extends Component<IProps> {
  public static navigationOptions = {
    title: "Add an employee",
  };

  public componentWillUnmount() {
    const { dispatchResetForm } = this.props;
    dispatchResetForm();
  }

  public render() {
    const {
      employeeName,
      phone,
      shift,
      dispatchCreateEmployee,
      createAction,
      navigation,
    } = this.props;
    return (
      <Card>
        <EmployeeForm {...this.props} asyncAction={createAction} />

        <CardSection>
          {renderButton(createAction.state === "PROGRESS", () =>
            dispatchCreateEmployee(
              {
                employeeName,
                phone,
                shift,
                uid: null,
              },
              navigation,
            ),
          )}
        </CardSection>
      </Card>
    );
  }
}

const mapStateToProps = (state: IRootState) => {
  const { employeeName, phone, shift, createAction } = state.employee;
  return {
    createAction,
    employeeName,
    phone,
    shift,
  };
};

export default connect(
  mapStateToProps,
  {
    dispatchCreateEmployee: createEmployee,
    dispatchFieldUpdate: updateField,
    dispatchResetForm: resetForm,
  },
)(CreateEmployee);
