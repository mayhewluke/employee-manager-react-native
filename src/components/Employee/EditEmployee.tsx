import React, { Component } from "react";
import { ScrollView, View } from "react-native";
import reactNativeCommunications from "react-native-communications";
import { NavigationScreenProp, NavigationScreenProps } from "react-navigation";
import { connect } from "react-redux";

import { AsyncValue, IRootState } from "../../store";
import {
  closeFireModal,
  fireEmployee,
  IEmployee,
  resetForm,
  ShiftDay,
  showFireModal,
  updateEmployee,
  updateField,
} from "../../store/Employee";
import { AsyncButton, Button, Card, CardSection, Confirm } from "../common";
import EmployeeForm, { IFormProps } from "./EmployeeForm";

interface IProps extends IFormProps, NavigationScreenProps {
  employeeName: string;
  phone: string;
  shift: ShiftDay;
  uid: string | null;
  fireAction: AsyncValue<null>;
  fireModalShown: boolean;
  updateAction: AsyncValue<null>;
  dispatchShowFireModal: () => void;
  dispatchUpdateEmployee: (employee: IEmployee<string>) => void;
  dispatchResetForm: () => void;
  dispatchCloseModal: () => void;
  dispatchFireEmployee: (
    uid: string | null,
    navigation: NavigationScreenProp<any>,
  ) => void;
}

const textSchedule = (phone: string, shift: ShiftDay) => () =>
  reactNativeCommunications.text(phone, `Your upcoming shift is on ${shift}`);

class EditEmployee extends Component<IProps> {
  public static navigationOptions = ({
    navigation,
  }: {
    navigation: NavigationScreenProp<any>;
  }) => ({
    title: `Edit ${navigation.getParam("employeeName")}`,
  });

  public componentWillUnmount() {
    const { dispatchResetForm } = this.props;
    dispatchResetForm();
  }

  public render() {
    const {
      employeeName,
      phone,
      shift,
      uid,
      fireAction,
      fireModalShown,
      updateAction,
      dispatchShowFireModal,
      dispatchUpdateEmployee,
      dispatchCloseModal,
      dispatchFireEmployee,
      navigation,
    } = this.props;
    // Short-circuit and redirect if the employee to edit has not actually been
    // saved yet (can't edit something non-persisted)
    if (uid === null) {
      navigation.navigate("EmployeeList");
      return null;
    }
    const onFireEmployee = () => dispatchFireEmployee(uid, navigation);
    const onUpdateEmployee = () =>
      dispatchUpdateEmployee({
        employeeName,
        phone,
        shift,
        uid,
      });
    return (
      <ScrollView>
        <Card>
          <EmployeeForm {...this.props} />

          <CardSection>
            <AsyncButton
              label="Save"
              asyncAction={updateAction}
              onPress={onUpdateEmployee}
            />
          </CardSection>

          <CardSection>
            <View style={{ flex: 1 }}>
              <Button onPress={textSchedule(phone, shift)}>
                Text schedule
              </Button>
            </View>
          </CardSection>

          <CardSection>
            <AsyncButton
              label="Fire"
              asyncAction={fireAction}
              onPress={dispatchShowFireModal}
            />
          </CardSection>
          <Confirm
            text={`Are you sure you want to fire ${employeeName}?`}
            isVisible={fireModalShown}
            onYes={onFireEmployee}
            onNo={dispatchCloseModal}
          />
        </Card>
      </ScrollView>
    );
  }
}

const mapStateToProps = (state: IRootState) => {
  const {
    employeeName,
    phone,
    shift,
    uid,
    fireModalShown,
    fireAction,
    updateAction,
  } = state.employee;
  return {
    employeeName,
    fireAction,
    fireModalShown,
    phone,
    shift,
    uid,
    updateAction,
  };
};

export default connect(
  mapStateToProps,
  {
    dispatchCloseModal: closeFireModal,
    dispatchFieldUpdate: updateField,
    dispatchFireEmployee: fireEmployee,
    dispatchResetForm: resetForm,
    dispatchShowFireModal: showFireModal,
    dispatchUpdateEmployee: updateEmployee,
  },
)(EditEmployee);
