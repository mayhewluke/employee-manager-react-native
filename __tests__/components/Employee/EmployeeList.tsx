import "jest-enzyme";
import React from "react";

import { shallow } from "enzyme";
import { TouchableOpacity } from "react-native";
import { Provider } from "react-redux";
import reactTestRenderer from "react-test-renderer";

import { stubNavigation } from "../../helpers/react-navigation";
import { StubbedStore } from "../../helpers/redux";

import EmployeeList from "../../../src/components/Employee/EmployeeList";
import { editEmployee, ShiftDay } from "../../../src/store/Employee";
import {
  IEmployeesState,
  unwatchEmployees,
  watchEmployees,
} from "../../../src/store/Employees";

jest.mock("../../../src/store/Employee");
jest.mock("../../../src/store/Employees");

// Otherwise TouchableOpacity does not get rendered properly for some reason,
// and does not have an onPress
jest.mock("TouchableOpacity", () => "TouchableOpacity");

describe("EmployeeList", () => {
  let props: any;
  let store: StubbedStore;
  let state: { employees: IEmployeesState };
  beforeEach(() => {
    state = {
      employees: { employeesAction: { state: "INIT" } },
    };
    store = new StubbedStore(state);
    props = {
      navigation: stubNavigation(),
    };
  });
  // This has side effects! Don't put this in beforeEach in case a test needs to
  // do something *before* running those side effects!
  const mkWrapper = () =>
    shallow(<EmployeeList {...props} />, {
      context: { store },
    }).dive();

  it("shows a spinner when data is loading", () => {
    state.employees = { employeesAction: { state: "PROGRESS" } };

    const wrapper = mkWrapper();

    expect(wrapper.find("Spinner").length).toBe(1);
  });

  it("does not show a spinner when data is not loading", () => {
    state.employees = { employeesAction: { state: "INIT" } }; // Good to be explicit

    const wrapper = mkWrapper();

    expect(wrapper.find("Spinner").length).toBe(0);
  });

  describe("with loaded Employees", () => {
    const foo = {
      employeeName: "foo",
      phone: "1234",
      shift: ShiftDay.Monday,
      uid: "uid1",
    };
    const bar = { ...foo, employeeName: "bar", uid: "uid2" };
    let component: any;
    beforeEach(() => {
      state.employees = {
        employeesAction: {
          state: "COMPLETE",
          value: { foo, bar },
        },
      };
      component = reactTestRenderer.create(
        <Provider store={store}>
          <EmployeeList {...props} />
        </Provider>,
      );
    });
    it("displays loaded Employees", () => {
      // Using a snapshot appears to be the only reasonable way to test this at
      // this point, even though it results in a clumsily-large snapshot
      expect(component.toJSON()).toMatchSnapshot();
    });
    it("links each Employee to EditEmployee", () => {
      const { navigation } = props;
      const employeeComponents = component.root.findAllByType(TouchableOpacity);

      employeeComponents[0].props.onPress();
      expect(editEmployee).toHaveBeenCalledWith(foo, navigation);

      employeeComponents[1].props.onPress();
      expect(editEmployee).toHaveBeenCalledWith(bar, navigation);

      expect(editEmployee).toHaveBeenCalledTimes(2);
    });
  });

  it("adds a button for CreateEmployee to the header", () => {
    const { navigation } = props;
    const headerRightWrapped = shallow(
      EmployeeList.navigationOptions({ navigation }).headerRight,
    );
    expect(navigation.navigate).not.toHaveBeenCalled();

    headerRightWrapped.find("[onPress]").simulate("press");

    expect(navigation.navigate).toHaveBeenCalledWith("CreateEmployee");
  });

  it("sets the page title", () => {
    expect(EmployeeList.navigationOptions({}).title).toEqual("Employees");
  });

  it("watches the Employees data on mount", () => {
    expect(watchEmployees).not.toHaveBeenCalled();
    mkWrapper();
    expect(watchEmployees).toHaveBeenCalledWith(props.navigation);
  });

  it("clears the watch on Employees on umount", () => {
    const wrapper = mkWrapper();
    expect(unwatchEmployees).not.toHaveBeenCalled();

    wrapper.unmount();
    expect(unwatchEmployees).toHaveBeenCalledWith(props.navigation);
  });
});
