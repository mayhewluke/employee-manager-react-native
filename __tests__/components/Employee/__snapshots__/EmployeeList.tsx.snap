// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeList with loaded Employees displays loaded Employees 1`] = `
<View>
  <RCTScrollView
    data={
      Array [
        Object {
          "employeeName": "foo",
          "phone": "1234",
          "shift": "Monday",
          "uid": "uid1",
        },
        Object {
          "employeeName": "bar",
          "phone": "1234",
          "shift": "Monday",
          "uid": "uid2",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <TouchableOpacity
          onPress={[Function]}
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                  "borderBottomWidth": 1,
                  "borderColor": "#ddd",
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                  "padding": 5,
                  "position": "relative",
                },
                undefined,
              ]
            }
          >
            <Text>
              foo
            </Text>
          </View>
        </TouchableOpacity>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <TouchableOpacity
          onPress={[Function]}
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                  "borderBottomWidth": 1,
                  "borderColor": "#ddd",
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                  "padding": 5,
                  "position": "relative",
                },
                undefined,
              ]
            }
          >
            <Text>
              bar
            </Text>
          </View>
        </TouchableOpacity>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
